### PR TITLE
Add mapping editor

### PR DESCRIPTION
**Pull request by DrSmugleaf**
## About the PR
@Emisse 

(no cut and pasting selections yet that is forbidden technology)

## Media
Demonstration of the entities, tiles and decals dropdowns, decal settings, search bar, pick and erase modes:

https://github.com/CM-14/CM-14/assets/10968691/498721b1-ba3f-4a31-b0f1-ca146934648a

---

Saving by pressing CTRL+S:

https://github.com/CM-14/CM-14/assets/10968691/4a06eac7-b9dd-4bc2-9969-c66f5a2b4e52

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
- Mapping actions have been replaced with a mapping editor.
- The `loadmapacts` command has been removed.

### CODE REVIEW

**Comment by mirrorcult** [comment]
on god?

**Comment by DrSmugleaf** [comment]
> on god?

on god

**Comment by MACMAN2003** [comment]
for real?

**Comment by DrSmugleaf** [comment]
Real reason I upstreamed this is so people's awful UI code stops breaking the mapping editor in CM14

**Comment by lzk228** [comment]
is it toggleable?
like by cvar

File: Content.Client/Mapping/MappingPrototype.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using Robust.Shared.Prototypes;
+
+namespace Content.Client.Mapping;
+
+public sealed class MappingPrototype
+{
```
Comment (metalgearsloth) [inline Content.Client/Mapping/MappingPrototype.cs:6]: Doc comment I beg.

File: Content.Client/Mapping/MappingPrototypeList.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        for (var i = Math.Min(prevStart - 1, endIndex); i >= startIndex; i--)
+        {
+            Insert(SearchList, _search[i], false).SetPositionInParent(0);
+        }
+
+        for (var i = Math.Max(prevEnd + 1, startIndex); i <= endIndex; i++)
+        {
+            Insert(SearchList, _search[i], false);
+        }
+    }
```
Comment (metalgearsloth) [inline Content.Client/Mapping/MappingPrototypeList.xaml.cs:155]: I'm stupid so comment would be nice here too.

File: Content.Client/Mapping/MappingState.cs [hunk 3]
```diff
@@ -0,0 +1,860 @@
...
+        _loadController.UnloadScreen();
+        UserInterfaceManager.UnloadScreen();
+        _input.Contexts.GetContext("common").RemoveFunction(ContentKeyFunctions.MappingUnselect);
+        _input.Contexts.GetContext("common").RemoveFunction(ContentKeyFunctions.SaveMap);
+        _input.Contexts.GetContext("common").RemoveFunction(ContentKeyFunctions.MappingEnablePick);
+        _input.Contexts.GetContext("common").RemoveFunction(ContentKeyFunctions.MappingEnableDelete);
+        _input.Contexts.GetContext("common").RemoveFunction(ContentKeyFunctions.MappingPick);
+        _input.Contexts.GetContext("common").RemoveFunction(ContentKeyFunctions.MappingRemoveDecal);
+        _input.Contexts.GetContext("common").RemoveFunction(ContentKeyFunctions.MappingCancelEraseDecal);
+        _input.Contexts.GetContext("common").RemoveFunction(ContentKeyFunctions.MappingOpenContextMenu);
```
Comment (metalgearsloth) [inline Content.Client/Mapping/MappingState.cs]: Should this just have its own input context that inherits common instead?

Comment (DrSmugleaf) [reply Content.Client/Mapping/MappingState.cs]: It would be possible but then you would need to change the context when entering/exiting the mapping state.

File: Content.Client/Mapping/MappingPrototypeList.xaml.cs (hunk 2 above)
Comment (DrSmugleaf) [reply Content.Client/Mapping/MappingPrototypeList.xaml.cs]: Done in [`a6c0bb4` (#23427)](https://github.com/space-wizards/space-station-14/pull/23427/commits/a6c0bb468d5457b3ee742c1b8363940bb1660cd0)

File: Content.Client/Mapping/MappingPrototype.cs (hunk 1 above)
Comment (DrSmugleaf) [reply Content.Client/Mapping/MappingPrototype.cs]: Done in [`a6c0bb4` (#23427)](https://github.com/space-wizards/space-station-14/pull/23427/commits/a6c0bb468d5457b3ee742c1b8363940bb1660cd0)

**Comment by DrSmugleaf** [comment]
> is it toggleable? like by cvar

It is activated by running the `mapping` command. You can also run the `scene MappingState` command to enable it and `scene GameplayState` command to disable it.

**Review by metalgearsloth — APPROVED** [review]

File: Content.Client/Mapping/MappingState.cs (hunk 3 above)
Comment (metalgearsloth) [reply Content.Client/Mapping/MappingState.cs]: Does that affect aghost or anything.

Comment (DrSmugleaf) [reply Content.Client/Mapping/MappingState.cs]: I don't think so, but it would conflict with the "editor" context of the F5/F6 windows.

File: Content.Client/Mapping/MappingState.cs [hunk 4]
```diff
@@ -0,0 +1,864 @@
...
+        _prototypes.Add(decals);
+
+        foreach (var decal in _prototypeManager.EnumeratePrototypes<DecalPrototype>())
+        {
+            Register(decal, decal.ID, decals);
+        }
+
+        Sort(mappings, decals);
+        mappings.Clear();
+    }
```
Comment (AJCM-git) [inline Content.Client/Mapping/MappingState.cs:226]: Localize the name, also It's very cursed and hard to follow through the way these "Parent" buttons are laid out, they should prob be their own explicit control instead of being implicitly differentiated by whether or not the prototype is null

**Review by AJCM-git — CHANGES_REQUESTED** [review]

**Review by AJCM-git — DISMISSED** [review]
Also i don't think this handles prototype reload? could be wrong tho, it either needs a easy way to reload the state or handle prototype reload directly

**Comment by Aexxie** [comment]
wow you really did it omg oh boy

**Comment by Kit0vras** [comment]
if I log into someone else's local server and save the map via CTRL+S, where will it go - to his directory or to my computer? in theory, I don't have access to his files, so quick save will only work for him, right?

**Comment by mirrorcult** [comment]
to your computer, the server serializes the map and sends it to the client

**Comment by DrSmugleaf** [comment]
Assuming that FULL_RELEASE isn't enabled, saving is currently disabled on full release though I could remove that condition

**Comment by Kit0vras** [comment]
to be honest, the new decal menu does not look more convenient (over the past year I have rarely used the decal ID search, relying on the visual component of the decal). Maybe, if possible, it should be made more similar to the old menu - a compact layout of decals with scrolling.
image
because a lot of scrolling through menus like decals in search of the right one causes a lot of pain, because decals focus on the visual component

**Comment by Peptide90** [comment]
> to be honest, the new decal menu does not look more convenient (over the past year I have rarely used the decal ID search, relying on the visual component of the decal). Maybe, if possible, it should be made more similar to the old menu - a compact layout of decals with scrolling.
> 
> image
> 
> because a lot of scrolling through menus like decals in search of the right one causes a lot of pain, because decals focus on the visual component

This is what I came here to check (what it's actually searching / categorising by).

IMO search should support not only ID but also check name at a minimum, at best also checking description and tags for matches. 

I still need to check how it's categorising but seeing Base often I assume it's by parent or ID. Again having more control of how to categorise stuff (by folder (path) for example or tags specific to mapping). This might help with decal menu if you could group stuff into alphabet, numerical, symbols and as we want for certain topics e.g. roads markings, flora, filth.

Otherwise ace PR and a long time coming.

**Comment by potato1234x** [comment]
holy shit best pr of the year and its only january

**Comment by FairlySadPanda** [comment]
How come this PR is withering on the vine?

**Comment by QuietlyWhisper** [comment]
For two reasons really, one, smug hasnt had any time to spend on this, and secondly she made an alternative tool for the purpose this was going to serve on cm14, so we haven't spent any more time on it since it's original creation

File: Content.Client/Mapping/MappingState.cs (hunk 3 above)
Comment (PJB3005) [reply Content.Client/Mapping/MappingState.cs]: This code makes me weep but I'll put it on the pile of things to fix when we nuke placement manager.

**Comment by TheShuEd** [comment]
Just a week away

File: Content.Client/Mapping/MappingState.cs (hunk 4 above)
Comment (metalgearsloth) [reply Content.Client/Mapping/MappingState.cs:226]: In this case the issue that a child can have any other node as its parent so the entire tree is the same type of mappingprototype, if it were just like a header I'd be more inclined to agree.
